shanoir-ng-front/src/app/shared/mass-download/mass-download.service.ts

```diff
@@ -0,0 +1,492 @@
+ * (at your option) any later version.
+ * 
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see https://www.gnu.org/licenses/gpl-3.0.html
+ */
+
+import { HttpResponse } from '@angular/common/http';
+import { ComponentRef, Injectable } from '@angular/core';
+import { Observable, Subject, Subscription } from 'rxjs-compat';
+import { Task, TaskStatus } from 'src/app/async-tasks/task.model';
+import { Dataset } from 'src/app/datasets/shared/dataset.model';
+import { DatasetService, Format } from 'src/app/datasets/shared/dataset.service';
+import { ServiceLocator } from 'src/app/utils/locator.service';
+import { NotificationsService } from '../notifications/notifications.service';
+import { DownloadSetupComponent } from './download-setup/download-setup.component';
+import { ConfirmDialogService } from '../components/confirm-dialog/confirm-dialog.service';
+import { Queue } from './queue.model';
+import { take } from 'rxjs/operators';
+import { SuperPromise } from 'src/app/utils/super-promise';
+import { DownloadSetupAltComponent } from './download-setup-alt/download-setup-alt.component';
+import { ConsoleService } from '../console/console.service';
+
+declare var JSZip: any;
+
+export type Report = {
+    taskId: number,
+    folderName: string,
+    requestedDatasetIds: number[],
+    startTime: number,
+    list?: {
+        [key: number]: {
+            status: 'QUEUED' | 'ERROR' | 'SUCCESS',
+            error?: any,
+            errorTime?: number
+        }
+    }
+    nbSuccess?: number;
+    nbError?: number;
+    duration?: number;
+    format: Format;
+    nbQueues: number;
+};
+
+@Injectable()
+export class MassDownloadService {
+
+    private downloadQueue: Queue = new Queue();
+    readonly BROWSER_COMPAT_ERROR_MSG: string = 'browser not compatible';
+    readonly REPORT_FILENAME: string = 'downloadReport.json';
+
+    constructor(
+        private datasetService: DatasetService,
+        private notificationService: NotificationsService,
+        private consoleService: ConsoleService,
+        private dialogService: ConfirmDialogService) {
+    }
+
+    downloadByIds(datasetIds: number[]): Promise<void> {
+        return this.openModal().then(ret => {
+            if (ret != 'cancel') {
+                return this._downloadByIds(datasetIds, ret.format, ret.nbQueues)
+            } else return Promise.resolve();
+        }).catch(error => {
+            if (error == this.BROWSER_COMPAT_ERROR_MSG) {
+                return this.openAltModal().then(ret => {
+                    if (ret != 'cancel') {
+                        return this._downloadByIdsAlt(datasetIds, ret);
+                    } else return Promise.resolve();
+                });
+            } else throw error;
+        });
+    }
+
+    downloadDatasets(datasets: Dataset[]): Promise<void> {
+        return this.openModal().then(ret => {
+            if (ret != 'cancel') {
```

Comment:
Why don't we test for the if (error == this.BROWSER_COMPAT_ERROR_MSG)  message for the next methods ?

Reply:
I think we also have to check if the user refuses the security check -> Then download normally ?

Reply:
Locally, I could not download datasets in ~/Downloads as it told me it could contain configuration files.

Reply:
* I added the checks